### ISSUE TITLE
[Sprint] Prevent Volatility Leakage in KNN Outlier Detection and Scaling

### ISSUE THREAD

**Issue by ankit02327** [ankit02327/stock-price#114]
As part of the Algorithmic & ML Sprint, we need to address severe data leakage and distance-metric vulnerabilities within the K-Nearest Neighbors model (`backend/algorithms/optimised/knn/knn.py`).

Currently, the `fit()` method calculates the mean and standard deviation of the entire feature matrix (`X`) to remove outliers that exceed 10 standard deviations. Because this is executed before any train/test splits, future market volatility directly influences which past data points are dropped—a massive data leakage vector. Furthermore, KNN relies heavily on geometric distance; the current use of `StandardScaler` distorts these distances when extreme financial outliers are present.

**Tasks:**
- [ ] Refactor the outlier detection logic in `fit()` to ensure statistics (mean/std) are calculated strictly on the training set, or transition to a rolling-window statistical filter.
- [ ] Replace `StandardScaler` with `sklearn.preprocessing.RobustScaler` to preserve distance integrity against extreme market spikes.
- [ ] Replace standard K-Fold CV with `TimeSeriesSplit` inside both `find_optimal_k` and `hyperparameter_tuning`.

**Comment by ankit02327** [comment]
1. **Isolate Outlier Filtering:** Instead of calculating `feature_std` and `feature_mean` on the entire `X_clean` matrix inside `fit()`, either remove the manual outlier logic in favor of `RobustScaler`, or implement a `sklearn.pipeline.Pipeline` with a custom transformer that calculates bounds exclusively during the `fit_transform` stage.
2. **Upgrade Scaler:** Swap `StandardScaler` for `RobustScaler` in `fit()`, `find_optimal_k()`, and `hyperparameter_tuning()`.
3. **Time-Aware CV:** Import `TimeSeriesSplit`. In `find_optimal_k()`, pass `cv=TimeSeriesSplit(n_splits=cv)` to `cross_val_score`. Repeat this for `GridSearchCV` in `hyperparameter_tuning()`.

**Comment by ankit02327** [comment]
```py
# 1. & 2. Robust Outlier Scaling (replaces manual std-dev outlier dropping)
from sklearn.preprocessing import RobustScaler
from sklearn.neighbors import KNeighborsRegressor

def fit(self, X: np.ndarray, y: np.ndarray) -> 'ModelInterface':
    self.validate_input(X, y)
    
    # Basic cleaning only (inf/nan) - remove the 10-std-dev manual filtering
    X_clean = np.nan_to_num(X, nan=0.0, posinf=1e6, neginf=-1e6)
    y_clean = np.nan_to_num(y, nan=np.nanmean(y), posinf=np.nanmax(y), neginf=np.nanmin(y))
    
    # ... valid mask filtering ...
    
    # Use RobustScaler which naturally handles outliers without data leakage
    self.scaler = RobustScaler()
    X_scaled = self.scaler.fit_transform(X_clean)
    
    self.model = KNeighborsRegressor(
        n_neighbors=self.n_neighbors, weights=self.weights,
        algorithm=self.algorithm, metric=self.metric, p=self.p, n_jobs=-1
    )
    self.model.fit(X_scaled, y_clean)
    # ... metrics calculation
    
# 3. Fix CV in find_optimal_k
from sklearn.model_selection import cross_val_score, TimeSeriesSplit

def find_optimal_k(self, X: np.ndarray, y: np.ndarray, k_range: List[int] = None, cv: int = 5) -> int:
    # ... setup ...
    tscv = TimeSeriesSplit(n_splits=cv)
    self.scaler = RobustScaler()
    X_scaled = self.scaler.fit_transform(X)
    
    for k in k_range:
        knn = KNeighborsRegressor(n_neighbors=k, weights=self.weights)
        scores = cross_val_score(knn, X_scaled, y, cv=tscv, scoring='neg_mean_squared_error')
        # ... finish tracking best score
```
